File: ROID-Forum-Web-Client/Convert Client Code/GMGWebsite/web/Website/MainTopBarSection.dart
library MainSection;
import 'Website.dart';
import 'dart:html';
import 'Sections/GenericSection.dart';
//import 'package:css_animation/css_animation.dart';

class MainSection {
  Website website;
  DivElement mainDiv;
  DivElement titleDiv;
  DivElement sectionsBarDiv;
  DivElement sectionBarHighlight;
  List<DivElement> divSections = new List<DivElement>();
  List<double> widthPercentages = new List();
  List<double> positionPercentages = new List();
  double currentHighlightWidth = 0.0;
  double currentHighlightPosition = 0.0;
  double newHighlightWidth = 0.0;
  double newHighlightPosition = 0.0;
  double hightlightMoveSpeed = 0.4;
  MainSection(Website w) {
    website = w;
    makeMainDivComponent();
    makeGMGTitleComponent();
    makeSectionBarComponent();
    makeSectionBarHighlightComponent();
    addSections();
    website.websiteDiv.append(mainDiv);
  }
  
  void darkTheme() {
    titleDiv.style.color = "white";
    sectionsBarDiv.style.color = "white";
  }
  void lightTheme() {
    titleDiv.style.color = "Black";
    sectionsBarDiv.style.color = "Black"; 
  }
  
  void logic() {
    if (currentHighlightWidth != newHighlightWidth) {
      double movementSpeed = ((newHighlightWidth-currentHighlightWidth)/5.0).abs();
      if (movementSpeed < hightlightMoveSpeed) {movementSpeed = hightlightMoveSpeed;}
      if (currentHighlightWidth < newHighlightWidth) {
        currentHighlightWidth += movementSpeed;
      }
      if (currentHighlightWidth > newHighlightWidth) {
        currentHighlightWidth -= movementSpeed;
      }
      if ((currentHighlightWidth - newHighlightWidth).abs() < movementSpeed) {
        currentHighlightWidth = newHighlightWidth;
      }
      sectionBarHighlight.style.width = "${currentHighlightWidth}%";
    }
    if (currentHighlightPosition != newHighlightPosition) {
      double movementSpeed = ((newHighlightPosition-currentHighlightPosition)/5.0).abs();
            if (movementSpeed < hightlightMoveSpeed) {movementSpeed = hightlightMoveSpeed;}
      if (currentHighlightPosition < newHighlightPosition) {
        currentHighlightPosition += movementSpeed;
      }
      if (currentHighlightPosition > newHighlightPosition) {
        currentHighlightPosition -= movementSpeed;
      }
      if ((currentHighlightPosition - newHighlightPosition).abs() < movementSpeed) {
        currentHighlightPosition = newHighlightPosition;
      }
      sectionBarHighlight.style.left = "${currentHighlightPosition}%";
    }
  }
  void sectionClick(int section) {
    for (int i = 0; i < divSections.length; i++) {
      DivElement d = divSections[i];
      d.style.textDecoration = '';
    }
    divSections[section].style.textDecoration = 'underline';
    website.showView(section);
    newHighlightWidth = widthPercentages[section];
    newHighlightPosition = positionPercentages[section];
  }

  /*********** Create GUI Components *************/

  void makeMainDivComponent() {
    mainDiv = new DivElement();
    mainDiv.style
        ..position = "absolute"
        ..left = "0px"
        ..top = "0px"
        ..height = "92px"
        ..width = "calc(100% - 300px)";
    
  }
  void makeGMGTitleComponent() {
    //Add the GMG name to the main div
    titleDiv = new DivElement();
    titleDiv.style
        ..position = "absolute"
        ..left = "0px"
        ..top = "20px"
        ..width = "100%"
        ..fontSize = "15px"
        ..textAlign = "center";
    titleDiv.text = 'Game Maker\'s Garage';
    mainDiv.append(titleDiv);
  }
  void makeSectionBarComponent() {
    //Add the section bar
    sectionsBarDiv = new DivElement();
    sectionsBarDiv.style
        ..position = "absolute"
        ..width = "80%"
        ..left = "10%"
        ..top = "50px"
        ..height = "40px"
        ..borderTopStyle = "solid"
        ..borderBottomStyle = "solid"
        ..borderWidth = "1px"
        ..borderColor = "#DDDDDD"
        ..userSelect = "none";
    mainDiv.append(sectionsBarDiv);
  }
  void addSections() {
    //Add the sections to the bar
    int totalLetters = 0;
    for (int i = 0; i < website.sectionOrder.length; i++) {
      GenericSection section = website.sectionOrder[i];
      totalLetters += section.getDisplayName().length;
    }
    double p = 0.0;
    for (int i = 0; i < website.sectionOrder.length; i++) {
      GenericSection section = website.sectionOrder[i];
      String sectionName = section.getDisplayName();
      DivElement s = new DivElement();
      s.text = sectionName;
      double width = 100.0 / totalLetters * sectionName.length;
      widthPercentages.add(width);
      positionPercentages.add(p);
      s.style
          ..position = 'absolute'
          ..width = '${width}%'
          ..textAlign = 'center'
          ..height = '40px'
          ..top = '0px'
          ..left = '${p}%'
          ..lineHeight = '40px'
          ..textOverflow = 'ellipsis'
          ..whiteSpace = 'nowrap'
          ..overflow = 'hidden'
          ..cursor = 'pointer';
      s.onClick.listen((var d) {
        sectionClick(i);
      });
      p += width;
      if (i != 0) {
        s.style.borderLeft = 'solid';
        s.style.borderColor = '#DDDDDD';
        s.style.borderWidth = '1px';
      }
      sectionsBarDiv.append(s);
      divSections.add(s);
      
    }
    
  }
  void makeSectionBarHighlightComponent() {
    sectionBarHighlight = new DivElement();
    sectionBarHighlight.style
        ..position = "absolute"
        ..top = "0px"
        ..left = "0px"
        ..height = "100%"
        ..width = "0%"
        ..backgroundColor = "rgba(100,200,100,0.5)";
    sectionsBarDiv.append(sectionBarHighlight);
  }
}

File: ROID-Forum-Web-Client/src/Interface/MainTopBarSection.ts
import {AppController} from "../AppController";
import {Interface} from "../Utility/Interface";
import {Section} from "./CodingSection/Section";

export class MainTopBarSection {
    website: AppController;
    mainDiv: HTMLDivElement;
    titleDiv: HTMLDivElement;
    sectionsBarDiv: HTMLDivElement;
    sectionBarHighlight: HTMLDivElement;
    divSections: HTMLDivElement[] = [];
    widthPercentages: number[] = [];
    positionPercentages: number[] = [];
    currentHighlightWidth = 0.0;
    currentHighlightPosition = 0.0;
    newHighlightWidth = 0.0;
    newHighlightPosition = 0.0;
    hightlightMoveSpeed = 0.4;

    constructor(w: AppController) {
        this.website = w;
        this.makeMainDivComponent();
        this.makeGMGTitleComponent();
        this.makeSectionBarComponent();
        this.makeSectionBarHighlightComponent();
        this.addSections();
        this.website.websiteDiv.appendChild(this.mainDiv);
    }

    darkTheme() {
        this.titleDiv.style.color = "white";
        this.sectionsBarDiv.style.color = "white";
    }

    lightTheme() {
        this.titleDiv.style.color = "Black";
        this.sectionsBarDiv.style.color = "Black";
    }

    logic() {
        if (this.currentHighlightWidth != this.newHighlightWidth) {
            let movementSpeed = Math.abs((this.newHighlightWidth - this.currentHighlightWidth) / 5.0);
            if (movementSpeed < this.hightlightMoveSpeed) {
                movementSpeed = this.hightlightMoveSpeed;
            }
            if (this.currentHighlightWidth < this.newHighlightWidth) {
                this.currentHighlightWidth += movementSpeed;
            }
            if (this.currentHighlightWidth > this.newHighlightWidth) {
                this.currentHighlightWidth -= movementSpeed;
            }
            if (Math.abs(this.currentHighlightWidth - this.newHighlightWidth) < movementSpeed) {
                this.currentHighlightWidth = this.newHighlightWidth;
            }
            this.sectionBarHighlight.style.width = `${this.currentHighlightWidth}%`;
        }
        if (this.currentHighlightPosition != this.newHighlightPosition) {
            let movementSpeed = Math.abs((this.newHighlightPosition - this.currentHighlightPosition) / 5.0);
            if (movementSpeed < this.hightlightMoveSpeed) {
                movementSpeed = this.hightlightMoveSpeed;
            }
            if (this.currentHighlightPosition < this.newHighlightPosition) {
                this.currentHighlightPosition += movementSpeed;
            }
            if (this.currentHighlightPosition > this.newHighlightPosition) {
                this.currentHighlightPosition -= movementSpeed;
            }
            if (Math.abs(this.currentHighlightPosition - this.newHighlightPosition) < movementSpeed) {
                this.currentHighlightPosition = this.newHighlightPosition;
            }
            this.sectionBarHighlight.style.left = `${this.currentHighlightPosition}%`;
        }
    }

    sectionClick(section: number) {
        for (let i = 0; i < this.divSections.length; i++) {
            let d: HTMLDivElement = this.divSections[i];
            d.style.textDecoration = '';
        }
        this.divSections[section].style.textDecoration = 'underline';
        this.website.showView(section);
        this.newHighlightWidth = this.widthPercentages[section];
        this.newHighlightPosition = this.positionPercentages[section];
    }

    /*********** Create GUI Components *************/

    makeMainDivComponent() {
        this.mainDiv = Interface.Create({type: 'div'});
        this.mainDiv.style.position = "absolute";
        this.mainDiv.style.left = "0px";
        this.mainDiv.style.top = "0px";
        this.mainDiv.style.height = "92px";
        this.mainDiv.style.width = "calc(100% - 300px)";

    }

    makeGMGTitleComponent() {
        //Add the GMG name to the main div
        this.titleDiv = Interface.Create({type: 'div'});
        this.titleDiv.style.position = "absolute";
        this.titleDiv.style.left = "0px";
        this.titleDiv.style.top = "20px";
        this.titleDiv.style.width = "100%";
        this.titleDiv.style.fontSize = "15px";
        this.titleDiv.style.textAlign = "center";
        this.titleDiv.innerText = 'Game Maker\'s Garage';
        this.mainDiv.appendChild(this.titleDiv);
    }

    makeSectionBarComponent() {
        //Add the section bar
        this.sectionsBarDiv = Interface.Create({type: 'div'});
        this.sectionsBarDiv.style.position = "absolute";
        this.sectionsBarDiv.style.width = "80%";
        this.sectionsBarDiv.style.left = "10%";
        this.sectionsBarDiv.style.top = "50px";
        this.sectionsBarDiv.style.height = "40px";
        this.sectionsBarDiv.style.borderTopStyle = "solid";
        this.sectionsBarDiv.style.borderBottomStyle = "solid";
        this.sectionsBarDiv.style.borderWidth = "1px";
        this.sectionsBarDiv.style.borderColor = "#DDDDDD";
        this.sectionsBarDiv.style.userSelect = "none";
        this.mainDiv.appendChild(this.sectionsBarDiv);
    }

    addSections() {
        //Add the sections to the bar
        let totalLetters = 0;
        for (let i = 0; i < this.website.sectionOrder.length; i++) {
            let section: Section = this.website.sectionOrder[i];
            totalLetters += section.getDisplayName().length;
        }
        let p = 0.0;
        for (let i = 0; i < this.website.sectionOrder.length; i++) {
            let section: Section = this.website.sectionOrder[i];
            let sectionName: string = section.getDisplayName();
            let s: HTMLDivElement = Interface.Create({type: 'div'});
            s.innerText = sectionName;
            let width: number = 100.0 / totalLetters * sectionName.length;
            this.widthPercentages.push(width);
            this.positionPercentages.push(p);
            s.style.position = 'absolute';
            s.style.width = `${width}%`;
            s.style.textAlign = 'center';
            s.style.height = '40px';
            s.style.top = '0px';
            s.style.left = `${p}%`;
            s.style.lineHeight = '40px';
            s.style.textOverflow = 'ellipsis';
            s.style.whiteSpace = 'nowrap';
            s.style.overflow = 'hidden';
            s.style.cursor = 'pointer';
            s.onclick = () => {
                this.sectionClick(i);
            };
            p += width;
            if (i != 0) {
                s.style.borderLeft = 'solid';
                s.style.borderColor = '#DDDDDD';
                s.style.borderWidth = '1px';
            }
            this.sectionsBarDiv.appendChild(s);
            this.divSections.push(s);

        }
    }

    makeSectionBarHighlightComponent() {
        this.sectionBarHighlight = Interface.Create({type: 'div'});
        this.sectionBarHighlight.style.position = "absolute";
        this.sectionBarHighlight.style.top = "0px";
        this.sectionBarHighlight.style.left = "0px";
        this.sectionBarHighlight.style.height = "100%";
        this.sectionBarHighlight.style.width = "0%";
        this.sectionBarHighlight.style.backgroundColor = "rgba(100,200,100,0.5)";
        this.sectionsBarDiv.appendChild(this.sectionBarHighlight);
    }
}
